specs/networking/p2p-interface.md

```diff
@@ -0,0 +1,712 @@
+
+#### Interop
+
+Unaggregated attestations from all shards are sent to the `beacon_attestation` topic.
+
+#### Mainnet
+
+Shards are grouped into their own subnets (defined by a shard topic). The number of shard subnets is defined via `SHARD_SUBNET_COUNT` and the shard `shard_number % SHARD_SUBNET_COUNT` is assigned to the topic: `shard{shard_number % SHARD_SUBNET_COUNT}_beacon_attestation`. Unaggregated attestations are sent to the subnet topic. Aggregated attestations are sent to the `beacon_attestation` topic.
+
+### Messages
+
+Each gossipsub [message](https://github.com/libp2p/go-libp2p-pubsub/blob/master/pb/rpc.proto#L17-L24) has a maximum size of `GOSSIP_MAX_SIZE`.
+
+Clients MUST reject (fail validation) messages that are over this size limit. Likewise, clients MUST NOT emit or propagate messages larger than this limit.
+
+The payload is carried in the `data` field of a gossipsub message, and varies depending on the topic:
+
+
+| Topic                        | Message Type      |
+|------------------------------|-------------------|
+| beacon_block                 | BeaconBlock       |
+| beacon_attestation           | Attestation       |
+| shard{N}\_beacon_attestation | Attestation       |
+| voluntary_exit               | VoluntaryExit     |
+| proposer_slashing            | ProposerSlashing  |
+| attester_slashing            | AttesterSlashing  |
+
+Clients MUST reject (fail validation) messages containing an incorrect type, or invalid payload.
+
+When processing incoming gossip, clients MAY descore or disconnect peers who fail to observe these constraints.
+
+### Encodings
+
+Topics are post-fixed with an encoding. Encodings define how the payload of a gossipsub message is encoded.
+
+#### Interop
+
+- `ssz` - All objects are SSZ-encoded. Example: The beacon block topic string is: `/beacon_block/ssz` and the data field of a gossipsub message is an ssz-encoded `BeaconBlock`.
+
+#### Mainnet
+
+- `ssz_snappy` - All objects are ssz-encoded and then compressed with snappy. Example: The beacon attestation topic string is: `/beacon_attestation/ssz_snappy` and the data field of a gossipsub message is an `Attestation` that has been ssz-encoded then compressed with snappy.
+
+Implementations MUST use a single encoding. Changing an encoding will require coordination between participating implementations.
+
+## The discovery domain: discv5
+
+Discovery Version 5 ([discv5](https://github.com/ethereum/devp2p/blob/master/discv5/discv5.md)) is used for peer discovery, both in the interoperability testnet and mainnet.
+
+`discv5` is a standalone protocol, running on UDP on a dedicated port, meant for peer discovery only. `discv5` supports self-certified, flexible peer records (ENRs) and topic-based advertisement, both of which are (or will be) requirements in this context.
```

Comment:
We should define what this dedicated port is ahead of time. Maybe have different ports for interop and mainnet

Reply:
Not necessary to define fixed ports. You only need to know the details of the bootstrappers. Past that, nodes announce their UDP ports in their self-certified ENR.

Reply:
Ahh right! ENRs have a field for both TCP and UDP ports.